web/src/components/Md.js

```diff
@@ -3,12 +3,89 @@ import React from 'react';
 
 import md from '../util/md';
 
-const Md = ({ content, wrapper: Wrapper, ...rest }) => (
-  <Wrapper
-    dangerouslySetInnerHTML={{ __html: md.renderInline(content) }}
-    {...rest}
-  />
-);
+const listTags = {
+  ol: 'ol',
+  ul: 'ul'
+};
+
+const Md = ({ content, wrapper: Wrapper, ...rest }) => {
+  // Our Markdown library accepts HTML content, but it gets all muddled up
+  // when there are linebreaks within the HTML.  Eg:
+  //
+  // <ul>
+  //   <li>one</li>
+  //   <li>two</li>
+  //   <li>three</li>
+  // </ul>
+  //
+  // ...ends up rendered as:
+  //
+  // <ul>
+  //   <br>
+  //   <li>one</li>
+  //   <br>
+  //   <li>two</li>
+  //   <br>
+  //   <li>three</li>
+  //   <br>
+  //  </ul>
+  //
+  // To workaround this, we can use some custom format tokens, such as [ul],
+  // to mark where lists should be.  Then we can parse out the content, build
+  // an HTML string with no extraneous whitespace or newlines, and put that
+  // back into the content before passing it on to the Markdown library.
+  //
+  // Declare this regex locally since we will be using exec(), which relies on
+  // internal regex state.  We don't want previous calls to impact this one!
```

Comment:
`internal regex state` is the most terrifying thing I've heard in recent memory. (the comments below explaining the regexes are beautiful, however.)